### ISSUE TITLE
Difference between "manual" hf-transformers embedding and sentence-transformers

### ISSUE THREAD

**Issue by buhrmann** [UKPLab/sentence-transformers#2569]
Hi, I noticed that when I manually extract embeddings from a Bert model using transformers with mean pooling, the individual embeddings change depending on the batch size. This is probably because huggingface's pipeline first orders the input texts by length, such that texts of more similar lengths are more probable to be processed in the same batch. And as a result, I imagine, the number of padding tokens is different for different batch sizes, which in turn influences the mean pooling.

Using sentence-transformers, on the other hand, I always get the same embeddings, independent of batch size, and these are equal to the embeddings I get with batch size = 1 using hf's transformers.

Is any of these two methods better than the other, i.e. would you recommend one over the other? I'd imagine hf's ordering means the padding tokens have less influence on the final embeddings, which in principle sounds reasonable? If that's the case, would it make sense to do the same in sentence-transformers?

**Comment by ir2718** [comment]
Hi,

sentence-transformers completely ignores the padding tokens during mean pooling (implemented [here](https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/models/Pooling.py#L138), lines 138 to 154), which is the reason for this difference. If you just use a mean operation over a batch of embeddings without taking into account the padding tokens you would get different results depending on the longest text in the batch, which is not something you usually want. For example, the way this is handled using raw transformers would be something like ([source](https://huggingface.co/sentence-transformers/all-mpnet-base-v2)):
```
def mean_pooling(model_output, attention_mask):
    token_embeddings = model_output[0] #First element of model_output contains all token embeddings
    input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
    return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(input_mask_expanded.sum(1), min=1e-9)
```

**Comment by buhrmann** [comment]
Thanks for the quick response! In that case it seems to be better and easier to just always use sentence-transformers instead of rolling one's own on top of HF's transformers. Nice!
